src/foam/dao/AbstractF3FileJournal.js

```diff
@@ -0,0 +1,485 @@
+/**
+ * @license
+ * Copyright 2019 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.dao',
+  name: 'AbstractF3FileJournal',
+  abstract: true,
+  flags: ['java'],
+
+  javaImports: [
+    'foam.core.FObject',
+    'foam.core.PropertyInfo',
+    'foam.core.ProxyX',
+    'foam.lib.formatter.FObjectFormatter',
+    'foam.lib.formatter.JSONFObjectFormatter',
+    'foam.lib.StoragePropertyPredicate',
+    'foam.lib.json.ExprParser',
+    'foam.lib.json.JSONParser',
+    'foam.lib.json.Outputter',
+    'foam.lib.parse.*',
+    'foam.nanos.auth.LastModifiedByAware',
+    'foam.nanos.auth.Subject',
+    'foam.nanos.auth.User',
+    'foam.nanos.logger.Logger',
+    'foam.nanos.logger.PrefixLogger',
+    'foam.nanos.logger.StdoutLogger',
+
+    'java.io.BufferedReader',
+    'java.io.BufferedWriter',
+    'java.io.InputStream',
+    'java.io.InputStreamReader',
+    'java.io.OutputStream',
+    'java.io.OutputStreamWriter',
+    'java.util.Calendar',
+    'java.util.Iterator',
+    'java.util.List',
+    'java.util.TimeZone',
+    'java.util.regex.Pattern'
+  ],
+
+  axioms: [
+    {
+      name: 'javaExtras',
+      buildJavaClass: function (cls) {
+        cls.extras.push(`
+          protected static Pattern COMMENT = Pattern.compile("(/\\\\*([^*]|[\\\\r\\\\n]|(\\\\*+([^*/]|[\\\\r\\\\n])))*\\\\*+/)|(//.*)");
+
+          protected static ThreadLocal<StringBuilder> sb = new ThreadLocal<StringBuilder>() {
```

Comment:
Better than having a thread-local StringBuilder is to have a thread-local Formatter so we can format in parallel.